README
NAME
    Data::GUID::Any - Generic interface for GUID/UUID creation

VERSION
    version 0.003

SYNOPSIS
        use Data::GUID::Any 'guid_as_string';

        my $guid = guid_as_string();

DESCRIPTION
    This module is a generic wrapper around various ways of obtaining
    Globally Unique ID's (GUID's), also known as Universally Unique
    Identifiers (UUID's).

    On installation, if Data::GUID::Any can't detect a way of generating
    both version 1 and version 4 GUID's, it will add either Data::UUID::MT
    or UUID::Tiny as a prerequisite, depending on whether or not a compiler
    is available.

USAGE
    The following functions are available for export.

  guid_as_string()
        my $guid = guid_as_string();

    Returns a guid in string format with upper-case hex characters:

      FA2D5B34-23DB-11DE-B548-0018F34EC37C

    This is the most general subroutine that offers the least amount of
    control over the result. This routine returns whatever is the default
    type of GUID for a source, which could be version 1 or version 4 (or, in
    the case of Win32, something resembling a version 1, but specific to
    Microsoft).

    It will use any of the following sources, listed from most preferred to
    least preferred:

    *   Data::UUID::MT (v4)

    *   Data::GUID (v1)

    *   Data::UUID (v1)

    *   Data::UUID::LibUUID (v4 or v1)

    *   UUID (v4 or v1)

    *   Win32 (using GuidGen()) (similar to v1)

    *   uuid (external program) (v1)

    *   APR::UUID (v4 or v1)

    *   UUID::Tiny (v1)

    At least one of them is guaranteed to exist or Data::GUID::Any will
    throw an exception when loaded. This shouldn't happen if prerequisites
    were correctly installed.

  v1_guid_as_string()
        my $guid = v1_guid_as_string();

    Returns a version 1 (timestamp+MAC/random-identifier) GUID in string
    format with upper-case hex characters from one of the following sources:

    *   Data::UUID::MT

    *   Data::GUID

    *   Data::UUID

    *   Data::UUID::LibUUID

    *   uuid (external program)

    *   UUID::Tiny

    If none of them are available, an exception will be thrown when this is
    called. This shouldn't happen if prerequisites were correctly installed.

  v4_guid_as_string()
        my $guid = v4_guid_as_string();

    Returns a version 4 (random) GUID in string format with upper-case hex
    characters from one of the following modules:

    *   Data::UUID::MT

    *   Data::UUID::LibUUID

    *   uuid (external program)

    *   UUID::Tiny

    If none of them are available, an exception will be thrown when this is
    called. This shouldn't happen if prerequisites were correctly installed.

SEE ALSO
    *   RFC 4122 [http://tools.ietf.org/html/rfc4122]

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Data-GUID-Any>. You
    will be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/data-guid-any>

      git clone https://github.com/dagolden/data-guid-any.git

AUTHOR
    David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Data-GUID-Any

0.003     2012-04-18 12:59:58 America/New_York
    
    - The ordering of GUID providers  has changed and there are now
      separate functions for generic GUID (any supported type), v1 and v4
      type GUIDs.
    
    - Data::UUID::MT is now preferred, since it offers good speed and the
      best control.

    - The uuid binary is now preferred to pure-perl UUID::Tiny.

    - UUID::Generator::PurePerl has been dropped as the author has listed
      it as deprecated.
    
    - distribution converted to Dist::Zilla style

0.002 Fri Apr 10 10:19:55 EDT 2009
    - added UUID::Generator::PurePerl
    - cleaned up some internals

0.001 Tue Apr  7 23:33:51 EDT 2009
    - initial release


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla.
# Don't edit it but the dist.ini used to construct it.
BEGIN { require 5.006; }
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;

my %WriteMakefileArgs = (
  'ABSTRACT' => 'Generic interface for GUID/UUID creation',
  'AUTHOR' => 'David Golden <dagolden@cpan.org>',
  'BUILD_REQUIRES' => {
    'Config' => '0',
    'File::Find' => '0',
    'File::Spec' => '0',
    'File::Temp' => '0',
    'Test::More' => '0.92'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::CBuilder' => '0.26',
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'Data-GUID-Any',
  'EXE_FILES' => [],
  'LICENSE' => 'apache',
  'NAME' => 'Data::GUID::Any',
  'PREREQ_PM' => {
    'Exporter' => '0',
    'IPC::Cmd' => '0',
    'strict' => '0',
    'warnings' => '0'
  },
  'VERSION' => '0.003',
  'test' => {
    'TESTS' => 't/*.t'
  }
);
;

$WriteMakefileArgs{PREREQ_PM} ||= {};

eval {
  local @INC = ('lib', @INC);
  require Data::GUID::Any;
  Data::GUID::Any::v1_guid_as_string(); # dies if no provider
  Data::GUID::Any::v4_guid_as_string(); # dies if no provider
  1;
} or do {
  require ExtUtils::CBuilder;
  if ( ExtUtils::CBuilder->new->have_compiler ) {
    $WriteMakefileArgs{PREREQ_PM}{'Data::UUID::MT'} = '0';
  }
  else {
    $WriteMakefileArgs{PREREQ_PM}{'UUID::Tiny'} = '0';
  }
};

# Hey, CPAN Testers, go ahead and test with extra prereqs
# that don't have external library dependencies
if ( $ENV{AUTOMATED_TESTING} ) {
  require ExtUtils::CBuilder;
  $WriteMakefileArgs{BUILD_REQUIRES}{'Data::UUID::MT'} = '0'
    if ExtUtils::CBuilder->new->have_compiler;
  $WriteMakefileArgs{BUILD_REQUIRES}{'UUID::Tiny'} = '0';
}
 


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



lib/Data/GUID/Any.pm
use 5.006;
use strict;
use warnings;
package Data::GUID::Any;
# ABSTRACT: Generic interface for GUID/UUID creation
our $VERSION = '0.003'; # VERSION

use IPC::Cmd;
use Exporter;
our @ISA = qw/Exporter/;
our @EXPORT_OK = qw/ guid_as_string v1_guid_as_string v4_guid_as_string/;

our ($Using_vX, $Using_v1, $Using_v4) = ("") x 3;

#--------------------------------------------------------------------------#

my $hex = "A-Z0-9";

sub _looks_like_guid {
  my $guid = shift;
  return $guid =~ /[$hex]{8}-[$hex]{4}-[$hex]{4}-[$hex]{4}-[$hex]{12}/;
}

#--------------------------------------------------------------------------#

# state variables for generator closures
my ($dumt_v1, $dumt_v4, $uuid_v1, $uuid_v4) = (undef) x 4; # reset if reloaded

my %generators = (
  # v1 or v4
  'Data::UUID::MT' => {
    type => 'module',
    v1 => sub {
      $dumt_v1 ||= Data::UUID::MT->new(version => 1);
      return uc $dumt_v1->create_string;
    },
    v4 => sub {
      $dumt_v4 ||= Data::UUID::MT->new(version => 4);
      return uc $dumt_v4->create_string;
    },
  },
  'Data::UUID::LibUUID' => {
    type => 'module',
    v1 => sub { return uc Data::UUID::LibUUID::new_uuid_string(2) },
    v4 => sub { return uc Data::UUID::LibUUID::new_uuid_string(4) },
    vX => sub { return uc Data::UUID::LibUUID::new_uuid_string() },
  },
  'UUID::Tiny' => {
    type => 'module',
    v1 => sub { return uc UUID::Tiny::create_UUID_as_string(UUID::Tiny::UUID_V1()) },
    v4 => sub { return uc UUID::Tiny::create_UUID_as_string(UUID::Tiny::UUID_V4()) },
  },
  'uuid' => {
    type => 'binary',
    v1 => sub {
      $uuid_v1 ||= IPC::Cmd::can_run('uuid');
      chomp( my $guid = qx/$uuid_v1 -v1/ ); return uc $guid;
    },
    v4 => sub {
      $uuid_v4 ||= IPC::Cmd::can_run('uuid');
      chomp( my $guid = qx/$uuid_v4 -v4/ ); return uc $guid;
    },
  },
  # v1 only
  'Data::GUID' => {
    type => 'module',
    v1 => sub { return uc Data::GUID->new->as_string },
  },
  'Data::UUID' => {
    type => 'module',
    v1 => sub { return uc Data::UUID->new->create_str },
  },
  # system dependent or custom
  'UUID' => {
    type => 'module',
    vX => sub { my ($u,$s); UUID::generate($u); UUID::unparse($u, $s); return uc $s },
  },
  'Win32' => {
    type => 'module',
    vX => sub { my $guid = Win32::GuidGen(); return uc substr($guid,1,-1) },
  },
  'APR::UUID' => {
    type => 'module',
    vX => sub { return uc APR::UUID->new->format },
  },
);

our $NO_BINARY; # for testing
sub _is_available {
  my ($name) = @_;
  if ( $generators{$name}{type} eq 'binary' ) {
    return $NO_BINARY ? undef : IPC::Cmd::can_run($name);
  }
  else {
    return eval "require $name";
  }
}

sub _best_generator {
  my ($list) = @_;
  for my $option ( @$list ) {
    my ($name, $version) = @$option;
    next unless my $g = $generators{$name};
    next unless _is_available($name);
    return ($name, $g->{$version})
      if $g->{$version} && _looks_like_guid( $g->{$version}->() );
  }
  return;
}

#--------------------------------------------------------------------------#

my %sets = (
  any => [
    ['Data::UUID::MT'       => 'v4'],
    ['Data::GUID'           => 'v1'],
    ['Data::UUID'           => 'v1'],
    ['Data::UUID::LibUUID'  => 'vX'],
    ['UUID'                 => 'vX'],
    ['Win32'                => 'vX'],
    ['uuid'                 => 'v1'],
    ['APR::UUID'            => 'vX'],
    ['UUID::Tiny'           => 'v1'],
  ],
  v1 => [
    ['Data::UUID::MT'       => 'v1'],
    ['Data::GUID'           => 'v1'],
    ['Data::UUID'           => 'v1'],
    ['Data::UUID::LibUUID'  => 'v1'],
    ['uuid'                 => 'v1'],
    ['UUID::Tiny'           => 'v1'],
  ],
  v4 => [
    ['Data::UUID::MT'       => 'v4'],
    ['Data::UUID::LibUUID'  => 'v4'],
    ['uuid'                 => 'v4'],
    ['UUID::Tiny'           => 'v4'],
  ],
);

sub _generator_set { return $sets{$_[0]} }

{
  no warnings qw/once redefine/;
  {
    my ($n, $s) = _best_generator(_generator_set("any"));
    die "Couldn't find a GUID provider" unless $n;
    *guid_as_string = $s;
    $Using_vX = $n;
  }
  {
    my ($n, $s) = _best_generator(_generator_set("v1"));
    *v1_guid_as_string = $s || sub { die "No v1 GUID provider found\n" };
    $Using_v1 = $n || '';
  }
  {
    my ($n, $s) = _best_generator(_generator_set("v4"));
    *v4_guid_as_string = $s || sub { die "No v4 GUID provider found\n" };
    $Using_v4 = $n || '';
  }
}

1;



=pod

=head1 NAME

Data::GUID::Any - Generic interface for GUID/UUID creation

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Data::GUID::Any 'guid_as_string';

    my $guid = guid_as_string();

=head1 DESCRIPTION

This module is a generic wrapper around various ways of obtaining Globally
Unique ID's (GUID's), also known as Universally Unique Identifiers (UUID's).

On installation, if Data::GUID::Any can't detect a way of generating both
version 1 and version 4 GUID's, it will add either Data::UUID::MT or UUID::Tiny
as a prerequisite, depending on whether or not a compiler is available.

=head1 USAGE

The following functions are available for export.

=head2 guid_as_string()

    my $guid = guid_as_string();

Returns a guid in string format with upper-case hex characters:

  FA2D5B34-23DB-11DE-B548-0018F34EC37C

This is the most general subroutine that offers the least amount of control
over the result.  This routine returns whatever is the default type of GUID for
a source, which could be version 1 or version 4 (or, in the case of Win32,
something resembling a version 1, but specific to Microsoft).

It will use any of the following sources, listed from most preferred to least
preferred:

=over 4

=item *

L<Data::UUID::MT> (v4)

=item *

L<Data::GUID> (v1)

=item *

L<Data::UUID> (v1)

=item *

L<Data::UUID::LibUUID> (v4 or v1)

=item *

L<UUID> (v4 or v1)

=item *

L<Win32> (using GuidGen()) (similar to v1)

=item *

uuid (external program) (v1)

=item *

L<APR::UUID> (v4 or v1)

=item *

L<UUID::Tiny> (v1)

=back

At least one of them is guaranteed to exist or Data::GUID::Any will
throw an exception when loaded. This shouldn't happen if prerequisites
were correctly installed.

=head2 v1_guid_as_string()

    my $guid = v1_guid_as_string();

Returns a version 1 (timestamp+MAC/random-identifier) GUID in string format
with upper-case hex characters from one of the following sources:

=over 4

=item *

L<Data::UUID::MT>

=item *

L<Data::GUID>

=item *

L<Data::UUID>

=item *

L<Data::UUID::LibUUID>

=item *

uuid (external program)

=item *

L<UUID::Tiny>

=back

If none of them are available, an exception will be thrown when this is called.
This shouldn't happen if prerequisites were correctly installed.

=head2 v4_guid_as_string()

    my $guid = v4_guid_as_string();

Returns a version 4 (random) GUID in string format with upper-case hex
characters from one of the following modules:

=over 4

=item *

L<Data::UUID::MT>

=item *

L<Data::UUID::LibUUID>

=item *

uuid (external program)

=item *

L<UUID::Tiny>

=back

If none of them are available, an exception will be thrown when this is called.
This shouldn't happen if prerequisites were correctly installed.

=head1 SEE ALSO

=over 4

=item *

RFC 4122 [http://tools.ietf.org/html/rfc4122]

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-GUID-Any>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/data-guid-any>

  git clone https://github.com/dagolden/data-guid-any.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


__END__


